File: 0189-rotate-array/0189-rotate-array.c

void reverseArray(int *nums, int start, int end) {
    if (start >= end) return;
    int temp = nums[start];
    nums[start] = nums[end];
    nums[end] = temp;
    reverseArray(nums, start + 1, end - 1);
}

void rotate(int* nums, int numsSize, int k){
    k = k % numsSize;
    reverseArray(nums, 0, numsSize - k - 1);
    reverseArray(nums, numsSize - k, numsSize-1);
    reverseArray(nums, 0, numsSize-1);
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def reverse_part(nums:List[int], start: int, end: int):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
        
        n = len(nums)
        steps = k % n
        reverse_part(nums, 0, n - steps - 1)
        reverse_part(nums, n - steps, n - 1)
        reverse_part(nums, 0, n -1)
        
